NVidia Tight encoder fix

maztheman opened the pull request:
Summary:

## fb.h
- Added additional field "byte_stride"
## fb.c
- Add byte stride to fb which is the width of a row in bytes (not pixels)
- fb size is no longer just using 4 for pixel size
## pixel.c
- added new function to handle 24 bit source format
## tight.c: 
- addressing into the source data will now use bits per pixel to calculate offsets
## damage-refinery.c
- use similar addressing formula as in tight.c, no longer is the bpp constant to 4.

any1 commented:
I'd rather see things be made more generic and based on bpp. Instead of using pixel32_to_cpixel and pixel24_to_cpixel inline, it would be better to have a pixel_to_cpixel that picks one based on the input pixel bpp.

We shouldn't continue indexing things as `uint32_t`. Instead, everything should be changed so that it uses `uint8_t` in `bpp` sized increments.

Byte stride is not a good name for "number of actual pixel bytes per row". Better stick to using bpp.

maztheman commented:
Do you know if there will be a performance impact switching everything to use `uint8_t` indexing? was that why it is being used currently?

any1 commented:
Everything is `uint32_t` for convenience.

maztheman commented:
latest change simplifies the code a lot.

any1 commented:
Maybe I should have told you that `pixel32_to_cpixel` is optimised. Plonking a memcpy in there like that is probably not going to do good things for it.

`bpp_stride` duplicates info that can be derived from other parameters in `nvnc_fb`.

maztheman commented:
address your latest comments.

any1 commented on src/zrle.c:
```diff
@@ -135,7 +135,7 @@ static void zrle_encode_packed_tile(struct vec* dst,
 
 	uint8_t cpalette[16 * 3];
-	pixel32_to_cpixel((uint8_t*)cpalette, dst_fmt, palette, src_fmt,
+	pixel_to_cpixel((uint8_t*)cpalette, dst_fmt, (uint8_t*)palette, src_fmt,
```
The type of `palette` should be changed rather than typecasting it here

maztheman replied:
I notice in copy tile 4 BPP is assumed, I think some modifications are necessary?

maztheman replied:
Also notice that the dst buffer size is calculated from the source but the output format could be different.

any1 replied:
Anything that assumes a static pixel size must be changed.

The destination buffer reserved size should be calculated based on the total number of pixels and the destination pixel size. It doesn't matter if the buffer is oversized, it just shouldn't be undersized. We only send what's written to the buffer.

maztheman replied:
ok, this makes sense, I will base the size of the destination format.

any1 replied:
This thread is still unresolved.

maztheman replied:
Once I started making this change it cascaded to something much larger and I think Id like to fix this file in a separate PR.

any1 commented on src/zrle.c:
```diff
@@ -196,7 +196,7 @@ static void zrle_encode_tile(struct vec* dst,
 	vec_fast_append_8(dst, 0);
 
-	pixel32_to_cpixel(((uint8_t*)dst->data) + 1, dst_fmt, src, src_fmt,
+	pixel_to_cpixel(((uint8_t*)dst->data) + 1, dst_fmt, (uint8_t*)src, src_fmt,
```
Ditto

maztheman commented on src/cursor.c:
```diff
@@ -94,12 +94,14 @@ int cursor_encode(struct vec* dst, struct rfb_pixel_format* pixfmt,
 	uint8_t* dstdata = dst->data;
 	dstdata += dst->len;
 
+	int32_t bpp_stride = image->width * (srcfmt.bits_per_pixel / 8);
```
Ok, i was not 100% sure what stride was defined as.  I will make these changes. Thanks for the input.

maztheman commented:
I'm still working on zrle.c but I also notice I had to patch wayvnc to pass proper stride.  if we are going to be using stride instead of width. I got zrle to not crash but the tiles look completely wrong.

any1 commented:
Pass the *proper* stride? Where is it failing to pass the stride along?

maztheman commented:
> Pass the _proper_ stride? Where is it failing to pass the stride along?

https://github.com/any1/wayvnc/blob/master/src/buffer.c#L96

maztheman commented:
im wodering if I can fix ZRLE in another MR as that will take some time to figure out what exactly is going on.  it is not crashing which is a plus but my current changes looks like this:

![image](https://github.com/any1/neatvnc/assets/1463024/31a3c2c7-246e-405e-949e-07a44f23e320)

any1 commented:
As long as zrle isn't broken for 32 bit pixel formats, it's fine.

maztheman commented:
Ok, changed to use stride, fixed some BPP issues that caused visual artifacts. I will create another PR for fixing zrle with Nvidia driver. 32 bit zrle is verified to still work.

any1 commented on src/cursor.c:
```diff
@@ -94,12 +94,14 @@ int cursor_encode(struct vec* dst, struct rfb_pixel_format* pixfmt,
 	uint8_t* dstdata = dst->data;
 	dstdata += dst->len;
 
+	int32_t bpp_stride = image->stride * (srcfmt.bits_per_pixel / 8);
```
This should be called `src_byte_stride`.

any1 commented on src/pixels.c:
```diff
@@ -215,6 +356,16 @@ int rfb_pixfmt_from_fourcc(struct rfb_pixel_format *dst, uint32_t src) {
 		dst->green_max = 0xff;
 		dst->blue_max = 0xff;
 		break;
+    case DRM_FORMAT_BGR888:
```
There's no RGB888?

any1 commented on src/tight.c:
```diff
@@ -335,6 +336,8 @@ static enum TJPF tight_get_jpeg_pixfmt(uint32_t fourcc)
 	case DRM_FORMAT_ABGR8888:
 	case DRM_FORMAT_XBGR8888:
 		return TJPF_RGBX;
+	case DRM_FORMAT_BGR888:
```
No RGB888?

any1 commented on src/pixels.c:
```diff
@@ -152,6 +151,148 @@ void pixel32_to_cpixel(uint8_t* restrict dst,
+		dst_endian_correction = dst_fmt->big_endian_flag ? 16 : 0;
+
+		while (len--) {
+			uint32_t cpx, px;
```
Here, for `bytes_per_pixel < 4`, `px` is partially uninitialised. By the way, there is a unit test for pixel conversion that you might find useful.

maztheman replied:
oh, I bet I also broke that unit test since pixel32 is static function now.

any1 replied:
`cpx = r | g | b`, not `px`. If you want to memcpy into it, you need to initialise it to 0 first.

any1 commented on test/test-pixels.c:
```diff
@@ -22,36 +22,74 @@
+
+static bool test_pixel_to_cpixel_3bpp(void)
+{
+	uint32_t src = u32_le(0x44112233u);
```
Would be good to mention that as a comment in the code itself.

any1 commented on src/pixels.c:
```diff
@@ -465,7 +625,8 @@ const char* drm_format_to_string(uint32_t fmt)
 	X(XBGR4444) \
-	X(RGB565)
+	X(RGB565)   \
+	X(BGR888)
```
Missing RGB888 here. Maybe it would be nicer to have them between 8888 and 4444 as these are listed in a descending order of bpp.

any1 commented on src/tight.c:
```diff
@@ -335,6 +336,9 @@ static enum TJPF tight_get_jpeg_pixfmt(uint32_t fourcc)
 	case DRM_FORMAT_XBGR8888:
 		return TJPF_RGBX;
+	case DRM_FORMAT_BGR888:
+	case DRM_FORMAT_RGB888:
```
Shouldn't `DRM_FORMAT_RGB888` yield `TJPF_BGR` and `DRM_FORMAT_BGR888` yield `TJPF_RGB`?

maztheman replied:
I think this conflicts with vnc clients...for instance tightvnc did not like jpeg BGR. I can make the change though if you want.

any1 replied:
That's odd. I understood this to be the format of the input buffer. The output format for JPEG is YCbCr, not RGB, so it's strange that this would affect the client. In what way did the client "not like" the output?

I find it surprising that your reds and blues aren't swapped because these format types are of opposite endianness.

maztheman replied:
hmm I wonder if the jpg should use the destination format, not source?

any1 replied:
https://rawcdn.githack.com/libjpeg-turbo/libjpeg-turbo/2.1.x/doc/html/group___turbo_j_p_e_g.html#gafbdce0112fd78fd38efae841443a9bcf

maztheman replied:
The colours are backwards. the nvidia card/driver i have the "source format" is BGR888. but when I use TJPF_BGR, the vnc on my phone the colours are backwards. now, this is probably an issue with the vnc client.

any1 replied:
DRM format is little-endian, but turbojpeg uses a big-endian format. This is why you need to return `TJPF_BGR` for `DRM_FORMAT_RGB888` and vice versa.

any1 commented on src/pixels.c:
```diff
@@ -457,6 +617,8 @@ const char* drm_format_to_string(uint32_t fmt)
 	X(XRGB8888) \
 	X(ABGR8888) \
 	X(XBGR8888) \
+	X(RGB888)   \
```
Nit: Don't align `\`

any1 commented on src/pixels.c:
```diff
@@ -275,6 +432,9 @@ int pixel_size_from_fourcc(uint32_t fourcc)
 		return 4;
+	case DRM_FORMAT_BGR888:
+	case DRM_FORMAT_RGB888:
+	    return 3;
```
Tabs, not spaces.

any1 reviewed (commented):
Seems to be taking shape. Please squash after making the requested changes.

any1 commented:
The squashed commit message shouldn't contain all the intermediate commit messages. That's just noise. It should not start with "enhancement:". That's reserved for saying *where* the changes are made not what sort of a change it is. See https://github.com/any1/wayvnc/blob/master/CONTRIBUTING.md#commit-messages, rule 8. As this commit has changes all over the place, this does not apply.

This commit does not *fully* implement 24 bit pixel formats, so it should say so. E.g. it might say something like "Implement 24 bit pixel formats for raw and tight". Then it is accurate.

It would have been best to implement this as a series of commits as follows, but I won't be too hard on you:
* pixels: Implement pixel24_to_cpixel
* fb: Use bpp in buffer size calculation
* damage-refinery: Implement arbitrary pixel sizes
* raw: Implement arbitrary pixel sizes
* right: Implement arbitrary pixel sizes

any1 commented on src/damage-refinery.c:
```diff
@@ -63,20 +64,22 @@ void damage_refinery_destroy(struct damage_refinery* self)
 	uint32_t hash = 0;
+	int32_t xoff = x_start * bpp;
 
 	// TODO: Support different pixel sizes
```
I guess this TODO can be removed now..

maztheman commented:
Thank you for reviewing this. I am very happy to get these changes in.

any1 commented:
LGTM. Can you go over outstanding threads and mark them as resolved, so that we can be sure that we're not forgetting anything?

maztheman commented:
Looks like everything was covered.

any1 commented:
I'll try to find the time to test this on the weekend.

any1 commented:
Thanks!
